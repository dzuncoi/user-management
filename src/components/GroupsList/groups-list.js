import React from 'react'
import PropTypes from 'prop-types'
import { Button, Table, Tooltip } from 'antd'
import { Link } from 'react-router-dom'

import Panel from '../Panel/panel'

import {
  NOT_LOADED,
  LOADING,
  SUCCEEDED,
  FAILED,
} from '../../constants/loading-status'

export const columns = [{
  title: 'Name',
  key: 'name',
  dataIndex: 'name',
  render: (val, group) => <Link to={`group/${group.id}`}>{val}</Link>,
}, {
  title: 'Users',
  key: 'users',
  render: (val, { users = [] }) => (users.length <= 0
    ? 'No Users'
    : (
      <Tooltip
        title={
          <React.Fragment>
            {users.map(user => <div key={user.id}>{user.firstName} {user.lastName}</div>)}
          </React.Fragment>
        }
      >
        {`${users.length} users`}
      </Tooltip>
    )
  ),
}, {
  title: 'Actions',
  key: 'actions',
  render: () => <Button type="ghost">Delete</Button>,
  fixed: 'right',
  width: 100,
}]

export const GroupsListComp = props => (
  <Panel
    title="Groups"
    extra={
      <Link to="group/new">
        <Button type="primary">New Group</Button>
      </Link>
    }
  >
    <Table
      name="groups"
      rowKey="id"
      dataSource={props.data}
      columns={columns}
      loading={props.loadingStatus === LOADING}
      scroll={{ x: true }}
    />
  </Panel>
)

GroupsListComp.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
  })),
  loadingStatus: PropTypes.oneOf([NOT_LOADED, LOADING, SUCCEEDED, FAILED]),
}

GroupsListComp.defaultProps = {
  data: [],
  loadingStatus: NOT_LOADED,
}

export default React.memo(GroupsListComp)
