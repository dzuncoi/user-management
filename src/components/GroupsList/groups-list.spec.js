import React from 'react'

import {
  NOT_LOADED,
  LOADING,
} from '../../constants/loading-status'
import { GroupsListComp, columns } from './groups-list'

jest.mock('antd', () => ({
  Button: () => 'Button',
  Table: () => 'Table',
  Tooltip: () => 'Tooltip',
}))
jest.mock('react-router-dom', () => ({
  Link: () => 'Link',
}))
jest.mock('../Panel/panel', () => 'Panel')

describe('groups-list', () => {
  let props

  beforeEach(() => {
    props = {
      loadingStatus: NOT_LOADED,
      data: [{
        id: '1',
        name: 'group 1',
      }],
      deleteGroup: jest.fn(),
    }
  })

  it('renders', () => {
    expect(shallow(<GroupsListComp {...props} />)).toMatchSnapshot()
  })

  it('renders loading on table if loadingStatus is LOADING', () => {
    const wrapper = shallow(<GroupsListComp {...props} loadingStatus={LOADING} />)
    expect(wrapper.find('Table').props().loading).toBeTruthy()
  })

  describe('table columns', () => {
    it('renders group name with link to edit group page', () => {
      const Col = columns[0].render('name', { id: 'id' })
      expect(Col.props.to).toEqual('group/id')
      expect(Col.props.children).toEqual('name')
    })

    it('renders No users if groups does not contain any users', () => {
      const Col = columns[1].render(null, { users: [] })
      expect(Col).toEqual('No Users')
    })

    it('renders amount of users belonging to this group', () => {
      const Col = columns[1].render(null, {
        users: [{
          id: '1',
        }, {
          id: '2',
        }],
      })
      expect(Col.props.children).toEqual('2 users')
    })
  })
})
