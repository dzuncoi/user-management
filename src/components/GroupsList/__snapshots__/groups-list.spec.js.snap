// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groups-list renders 1`] = `
<Panel
  extra={
    <Link
      to="group/new"
    >
      <Button
        type="primary"
      >
        New Group
      </Button>
    </Link>
  }
  title="Groups"
>
  <Table
    columns={
      Array [
        Object {
          "dataIndex": "name",
          "key": "name",
          "render": [Function],
          "title": "Name",
        },
        Object {
          "key": "users",
          "render": [Function],
          "title": "Users",
        },
        Object {
          "fixed": "right",
          "key": "actions",
          "render": [Function],
          "title": "Actions",
          "width": 100,
        },
      ]
    }
    dataSource={
      Array [
        Object {
          "id": "1",
          "name": "group 1",
        },
      ]
    }
    loading={false}
    name="groups"
    rowKey="id"
    scroll={
      Object {
        "x": true,
      }
    }
  />
</Panel>
`;
